test/unit/control-connection-tests.js

```diff
@@ -133,85 +138,60 @@ describe('ControlConnection', function () {
-      cc.nodeSchemaChangeHandler({ schemaChangeType: 'UPDATED', keyspace: ksName, isKeyspace: true});
-      assert.deepEqual(refreshedKeyspaces, [ ksName, ksName ]);
-      cc.nodeSchemaChangeHandler({ schemaChangeType: 'UPDATED', keyspace: ksName, isKeyspace: true});
-      assert.deepEqual(refreshedKeyspaces, [ ksName, ksName, ksName ]);
-      cc.metadata.keyspaces[ksName] = { tables: { 'tbl1': {} }, views: {} };
-      cc.nodeSchemaChangeHandler({ schemaChangeType: 'UPDATED', keyspace: ksName, table: 'tbl1'});
-      // clears the internal state
-      assert.ok(!cc.metadata.keyspaces[ksName].tables['tbl1']);
+    it('should continue iterating through the hosts when borrowing a connection fails', function (done) {
+      var hosts = [];
+      var cc = newInstance({ contactPoints: [ '::1', '::2' ] }, getContext({ hosts: hosts, failBorrow: [ 0 ] }));
+      cc.init(function (err) {
+        cc.shutdown();
+        assert.ifError(err);
+        assert.strictEqual(hosts.length, 2);
+        assert.ok(cc.initialized);
+        done();
+      });
     });
-  });
-  describe('#nodeStatusChangeHandler()', function () {
-    it('should call event address toString() to get', function () {
-      var options = clientOptions.extend({}, helper.baseOptions);
-      var toStringCalled = false;
-      var hostsGetCalled = false;
-      var cc = newInstance(options);
-      cc.hosts = { get : function () { hostsGetCalled = true;}};
-      var event = { inet: { address: { toString: function () { toStringCalled = true; return 'host1';}}}};
-      cc.nodeStatusChangeHandler(event);
-      assert.strictEqual(toStringCalled, true);
-      assert.strictEqual(hostsGetCalled, true);
+    it('should callback with NoHostAvailableError when borrowing all connections fail', function (done) {
+      var hosts = [];
+      var cc = newInstance({ contactPoints: [ '::1', '::2' ] }, getContext({ hosts: hosts, failBorrow: [ 0, 1] }));
+      cc.init(function (err) {
+        cc.shutdown();
+        helper.assertInstanceOf(err, errors.NoHostAvailableError);
+        assert.strictEqual(Object.keys(err.innerErrors).length, 2);
+        assert.strictEqual(hosts.length, 2);
+        assert.ok(!cc.initialized);
+        done();
+      });
     });
-    it('should set the node down when distance is ignored', function () {
-      var downSet = 0;
-      var options = clientOptions.extend({}, helper.baseOptions);
-      var cc = newInstance(options);
-      cc.hosts = { 
-        get : function () { 
-          return {
-            setDown: function () { downSet++; },
-            setDistance: function () { return types.distance.ignored; }
-          };
-        }
-      };
-      var event = { inet: { address: { toString: function () { return 'host1';}}}};
-      cc.nodeStatusChangeHandler(event);
-      assert.strictEqual(downSet, 1);
+    it('should continue iterating through the hosts when metadata retrieval fails', function (done) {
```

Comment:
Nice, will be difficult to test at integration level so nice to have a unit test that exercises the behavior.